### PR TITLE
Feature: Excel, keep table output of previous worksheet

### LINKED ISSUE #448: [Excel] Create new table + keep prev output

### PR DESCRIPTION
**Pull request by Etschbeijer**
Closes #448

### CODE REVIEW

**Comment by Etschbeijer** [comment]
1st table
345032464-0e4dd306-749c-437d-891e-266cbb91145a-1

2nd table
345033053-895583b8-ef8e-4f04-87da-2c7ebde8283e

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 1]
```diff
@@ -74,27 +100,32 @@ module OfficeInteropExtensions =
...
-            //                    stringCell
-            //            else stringCell
-            //        let address = FsAddress(rowI+1,colI+1)
-            //        fsTable.Cell(address, ws.CellCollection).SetValueAs value
-            //    )  
-            //)
-            //ws
-
-    let x = 0
+        static member arcTableFromExcelTableName (tableName:string, context:RequestContext) =
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:103]: name can be shortened `ArcTable.arcTableFromExcelTableName` duplicates arctable. Can be shortened to `ArcTable.fromExcelTableName` without loosing information

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 2]
```diff
@@ -206,30 +237,53 @@ let getPrevAnnotationTable (context:RequestContext) =
...
         if prevTableName.IsSome then
-            // Ref. 2
-            let! buildingBlocks = BuildingBlockFunctions.getBuildingBlocks context prevTableName.Value
+        
+            let! result = ArcTable.arcTableFromExcelTableName(prevTableName.Value, context)
+            return result
+
+        else
+
+            let! result = ArcTable.arcTableFromExcelTableName("", context)
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:252]: what does this logic do? Why do we get an arctable from ""-name. Shouldn't this be tryGet? With Some/None?

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 3]
```diff
@@ -314,26 +370,50 @@ let private createAnnotationTableAtRange (isDark:bool, tryUseLastOutput:bool, ra
...
+
+            //logic to compare size of previous table and current table and adapt size of inMemory table
             if useExistingPrevOutput then
-                let newColValues = prevTableOutput |> Array.map (fun cell -> ResizeArray[|Option.bind (box >> Some) cell.Value|] ) |> ResizeArray
-                let col1 = (annotationTable.columns.getItemAt 0.)
-                let body = col1.getDataBodyRange()
+                let rowCount0 = int table.rows.count
+                let diff = rowCount0 - newColValues.Count
+
+                if diff > 0 then // table larger than values -> Delete rows to reduce inMemoryTable size to previous table size
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:394]: We adjust exel table not in-memory table!

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 4]
```diff
@@ -314,26 +370,50 @@ let private createAnnotationTableAtRange (isDark:bool, tryUseLastOutput:bool, ra
...
             // Update annotationTable style
             annotationTable.style <- style
-    
+            annotationTable
+        )
+
+        let _ = table.rows.load(propertyNames = U2.Case2 (ResizeArray[|"count"|]))
+
+        //Skip header because it is newly generated for inMemory table
+        let newColValues =
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:379]: We compare values to rows.count, are we sure that rows.count excludes header row?

File: src/Client/OfficeInterop/OfficeInterop.fs [hunk 5]
```diff
@@ -314,26 +370,50 @@ let private createAnnotationTableAtRange (isDark:bool, tryUseLastOutput:bool, ra
...
+
             // Fit widths and heights of cols and rows to value size. (In this case the new column headers).
             activeSheet.getUsedRange().format.autofitColumns()
             activeSheet.getUsedRange().format.autofitRows()
     
             r.enableEvents <- true
     
             // Return info message
-            let logging = InteropLogging.Msg.create InteropLogging.Info (sprintf "Annotation Table created in [%s] with dimensions 2c x (%.0f + 1h)r." tableRange.address (tableRange.rowCount - 1.))
+            let logging = InteropLogging.Msg.create InteropLogging.Info (sprintf "Annotation Table created in [%s] with dimensions 2c x (%.0f + 1h)r." tableRange.address (6. - 1.))
```
Comment (Freymaurer) [inline src/Client/OfficeInterop/OfficeInterop.fs:411]: Why do we hardcode rowCount to 6?

**Review by Freymaurer — CHANGES_REQUESTED** [review]

**Review by Freymaurer — APPROVED** [review]

**Pull request merged**